<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# no-new-array

> [ESLint rule][eslint-rules] disallowing the use of the `new Array()` constructor.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-new-array' );
```

#### rule

[ESLint rule][eslint-rules] disallowing the use of the `new Array()` constructor.

**Bad**:

<!-- eslint-disable stdlib/no-new-array -->

```javascript
var arr = new Array( 5 );
var i;
for ( i = 0; i < 5; i++ ) {
    arr[ i ] = i;
}

console.log( arr );
// => [ 0, 1, 2, 3, 4 ]
```

**Good**:

```javascript
var arr = [];
var i;
for ( i = 0; i < 5; i++ ) {
    arr.push( i );
}

console.log( arr );
// => [ 0, 1, 2, 3, 4 ]
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-new-array' );

var linter = new Linter();

var code = 'var arr = new Array( 5 );';

linter.defineRule( 'no-new-array', rule );

var result = linter.verify( code, {
    'rules': {
        'no-new-array': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'no-new-array',
            'severity': 2,
            'message': 'Using the `new Array()` constructor is not allowed; use an array literal with push instead.',
            'line': 1,
            'column': 14,
            'nodeType': 'CallExpression',
            'source': 'var arr = new Array( 5 );',
            'endLine': 1,
            'endColumn': 25
        }
    ]
*/
```

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
