<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# no-unnecessary-nested-functions

> [ESLint rule][eslint-rules] to prevent unnecessary function nesting when inner functions don't depend on outer scope variables.

<section class="intro">

This rule enforces a best practice to only have functions defined inside of other functions if they need access to variables defined in the outer function scope. Otherwise, functions should always be elevated to the highest level (ideally module scope) to avoid redefining the functions per invocation.

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions' );
```

#### rule

[ESLint rule][eslint-rules] to prevent unnecessary function nesting when inner functions don't depend on outer scope variables.

**Bad**:

<!-- eslint-disable stdlib/no-unnecessary-nested-functions -->

```javascript
var PI = require( '@stdlib/constants/float64/pi' );

function outer( x ) {
    var result;

    function helper() {
        return PI; // doesn't use any variables from outer scope...
    }

    result = x * helper();
    return result;
}
```

**Good**:

```javascript
var PI = require( '@stdlib/constants/float64/pi' );

// Helper moved to module scope...
function helper() {
    return PI;
}

function outer( x ) {
    var result;
    result = x * helper();
    return result;
}
```

**Good** (function uses outer scope variables):

```javascript
function outer( x ) {
    var multiplier = 2;

    function helper() {
        return x * multiplier; // uses 'multiplier' from outer scope...
    }

    return helper();
}
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   The rule only checks function **declarations**, not function expressions or arrow functions.
-   Functions already at module or global scope are ignored.
-   Functions that reference variables from outer scopes are not flagged (closures are preserved).
-   Functions inside IIFEs (Immediately Invoked Function Expressions) are not flagged, as IIFEs are intentionally used for scope isolation.

</section>

<!-- /.notes -->

<section class="examples">

## Examples

<!-- eslint no-undef: "error" -->

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-unnecessary-nested-functions' );

var linter = new Linter();

// Generate source code containing nested function without dependencies:
var code = [
    'function outer() {',
    '  function inner() {',
    '    return 42;',
    '  }',
    '  return inner();',
    '}'
].join( '\n' );

// Define the ESLint configuration:
var config = {
    'rules': {
        'no-unnecessary-nested-functions': 'error'
    }
};

// Register the rule:
linter.defineRule( 'no-unnecessary-nested-functions', rule );

// Lint the code:
var out = linter.verify( code, config );
console.log( out );
/* =>
    [
        {
            'ruleId': 'no-unnecessary-nested-functions',
            'severity': 2,
            'message': 'Function \'inner\' should be moved to module scope.',
            'line': 2,
            'column': 3,
            'nodeType': 'FunctionDeclaration',
            'endLine': 4,
            'endColumn': 4
        }
    ]
*/
```

</section>

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
