<!--

@license Apache-2.0

Copyright (c) 2025 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# no-builtin-big-int

> [ESLint rule][eslint-rules] disallowing the use of the built-in global `BigInt` literal syntax and constructor.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var rule = require( '@stdlib/_tools/eslint/rules/no-builtin-big-int' );
```

#### rule

[ESLint rule][eslint-rules] disallowing the use of the built-in global `BigInt` literal syntax and constructor.

**Bad**:

<!-- eslint-disable stdlib/no-builtin-big-int -->

<!-- eslint-disable stdlib/require-globals -->

```javascript
var x = BigInt( 123 );
console.log( typeof x );
// => 'bigint'
```

<!-- eslint-enable stdlib/require-globals -->

**Good**:

```javascript
var BigInt = require( '@stdlib/bigint/ctor' );

var x = BigInt( 123 );
console.log( typeof x );
// => 'bigint'
```

</section>

<!-- /.usage -->

<section class="examples">

## Examples

```javascript
var Linter = require( 'eslint' ).Linter;
var rule = require( '@stdlib/_tools/eslint/rules/no-builtin-big-int' );

var linter = new Linter();

var code = 'var x = 123n;';

linter.defineRule( 'no-builtin-big-int', rule );

var result = linter.verify( code, {
    'rules': {
        'no-builtin-big-int': 'error'
    }
});
/* returns
    [
        {
            'ruleId': 'no-builtin-big-int',
            'severity': 2,
            'message': 'Using the built-in global `BigInt` literal syntax is not allowed; use `@stdlib/bigint/ctor` instead.',
            'line': 1,
            'column': 9,
            'nodeType': 'Literal',
            'source': 'var x = 123n;',
            'endLine': 1,
            'endColumn': 13
        }
    ]
*/
```

<!-- /.examples -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

<section class="related">

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[eslint-rules]: https://eslint.org/docs/developer-guide/working-with-rules

</section>

<!-- /.links -->
