// @stdlib/_tools/lint/pkg-json-names/test/test.sync.js
/**
* @license Apache-2.0
*
* Copyright (c) 2025 The Stdlib Authors.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

'use strict';

// MODULES //

var resolve = require( 'path' ).resolve;
var tape = require( 'tape' );
var isObjectArray = require( '@stdlib/assert/is-object-array' );
var isString = require( '@stdlib/assert/is-string' );
var isPlainObject = require( '@stdlib/assert/is-plain-object' );
var lintSync = require( './../lib' ).sync;


// TESTS //

tape( 'main export is a function', function test( t ) {
	t.ok( true, __filename );
	t.strictEqual( typeof lintSync, 'function', 'main export is a function' );
	t.end();
});

tape( 'the function throws an error if provided invalid options', function test( t ) {
	var values;
	var i;

	values = [
		'5',
		5,
		NaN,
		true,
		false,
		null,
		void 0,
		[],
		function noop() {}
	];

	for ( i = 0; i < values.length; i++ ) {
		t.throws( badValue( values[ i ] ), TypeError, 'throws a type error when provided ' + values[ i ] );
	}
	t.end();

	function badValue( value ) {
		return function badValue() {
			lintSync( value );
		};
	}
});

tape( 'the function returns `null` if unable to resolve any `package.json` files', function test( t ) {
	var opts = {
		'dir': './nonexistent',
		'pattern': '**/*.json'
	};
	var errs = lintSync( opts );
	t.strictEqual( errs, null, 'returns expected value' );
	t.end();
});

tape( 'the function returns `null` if unable to resolve any `package.json` files (empty options)', function test( t ) {
	var opts = {
		'dir': './nonexistent'
	};
	var errs = lintSync( opts );
	t.strictEqual( errs, null, 'returns expected value' );
	t.end();
});

tape( 'the function accepts options', function test( t ) {
	var opts;
	var errs;

	opts = {
		'dir': './nonexistent',
		'pattern': 'nothing.json'
	};

	errs = lintSync( opts );
	t.strictEqual( errs, null, 'returns expected value' );
	t.end();
});

tape( 'the function returns an array of lint errors if one or more files have validation errors', function test( t ) {
	var opts;
	var errs;

	opts = {
		'dir': resolve( __dirname, 'fixtures' ),
		'pattern': 'bad.json'
	};

	errs = lintSync( opts );
	t.strictEqual( isObjectArray( errs ), true, 'returns expected value' );
	t.strictEqual( errs.length, 1, 'returns expected value' );
	t.strictEqual( isString( errs[ 0 ].file ), true, 'returns expected value' );
	t.strictEqual( isPlainObject( errs[ 0 ].error ), true, 'returns expected value' );
	t.end();
});
